conda-store-server/conda_store_server/orm.py

```diff
@@ -234,15 +257,29 @@ def build_key(self):
         The build key should be a key that allows for the environment
         build to be easily identified and found in the database.
         """
-        datetime_format = "%Y%m%d-%H%M%S-%f"
-        return f"{self.specification.sha256}-{self.scheduled_on.strftime(datetime_format)}-{self.id}-{self.specification.name}"
+        if self.build_key_version == 1:
+            datetime_format = "%Y%m%d-%H%M%S-%f"
+            return f"{self.specification.sha256}-{self.scheduled_on.strftime(datetime_format)}-{self.id}-{self.specification.name}"
+        elif self.build_key_version == 2:
+            hash = self.specification.sha256[:_BUILD_KEY_V2_HASH_SIZE]
+            timestamp = int(self.scheduled_on.timestamp())
+            id = self.id
+            name = self.specification.name
+            return f"{hash}-{timestamp}-{id}-{name}"
+        else:
+            raise ValueError(f"invalid build key version: {self.build_key_version}")
 
     @staticmethod
     def parse_build_key(key):
         parts = key.split("-")
-        if len(parts) < 5:
-            return None
-        return int(parts[4])  # build_id
+        # Note: cannot rely on the number of dashes to differentiate between
+        # versions because name can contain dashes. Instead, this relies on the
+        # hash size to infer the format. The name is the last field, so indexing
+        # to find the id is okay.
+        if key[_BUILD_KEY_V2_HASH_SIZE] == "-":  # v2
+            return int(parts[2])  # build_id
+        else:  # v1
+            return int(parts[4])  # build_id
```

Comment:
This is also a bit hacky. We shouldn't rely on lossy info for this, but a well-known id->info relationship in the database. I know you know, and that that'd be considered ideal and we don't have the time but I wonder what's the rush to ship this now instead of in two weeks (still within the STF window).

If this is not going to be addressed now, and if we don't have an issue for this limitation, we should create one.

Reply:
I want to ship this version of the build key regardless because:
- it somewhat mitigates the length issues people might be having (see the test, the new version is 2x shorter)
- it's close enough to what we have already so risk of potential issues is minimized
- I consider Nix-like hashes a riskier change and would like to have this as an intermediate step, so people could fall back to this if needed.